### PR TITLE
feat: v4 non-complex hooks routing

### PR DESCRIPTION
**Pull request by jsy1218**
we can support non-complex hooks routing, that we know have no effect during swaps.

### CODE REVIEW

**Comment by jsy1218** [comment]
* **#916** <a href="https://app.graphite.dev/github/pr/Uniswap/routing-api/916?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Uniswap/routing-api/916?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: test/utils/v4HooksPoolsFiltering.ts [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    !Hook.hasSwapPermissions(pool.hooks) &&
+    // If the fee tier is smaller than or equal to 100%, it means the pool is not dynamic fee pool.
+    // Swap fee in total can be 100% (https://github.com/Uniswap/v4-core/blob/b619b6718e31aa5b4fa0286520c455ceb950276d/src/libraries/SwapMath.sol#L12)
+    // Dynamic fee is at 0x800000 or 838.8608% fee tier.
+    // Since pool manager doesn;t check the fee at 100% max during pool initialization (https://github.com/Uniswap/v4-core/blob/main/src/PoolManager.sol#L128)
+    // it's more defensive programming to ensure the fee tier is less than or equal to 100%
+    Number(pool.feeTier) <= 1000000
+  )
+}
+
```
Comment (jsy1218) [inline test/utils/v4HooksPoolsFiltering.ts:17]: pending unit tests

File: test/utils/v4HooksPoolsFiltering.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      const v4Pools = v4PoolsByTokenPairsAndFees[pool.token0.id.concat(pool.token1.id).concat(pool.feeTier)] || []
+
+      v4Pools.push(pool)
+      v4Pools.sort((a, b) => (a.tvlETH > b.tvlETH ? -1 : 1))
+
+      if (v4Pools.length > 10) {
+        v4Pools.pop()
+      }
+
+      v4PoolsByTokenPairsAndFees[pool.token0.id.concat(pool.token1.id).concat(pool.feeTier)] = v4Pools
```
Comment (xrsv) [inline test/utils/v4HooksPoolsFiltering.ts:35]: better to create a `createV4PoolKey(pool)`

File: test/utils/v4HooksPoolsFiltering.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+  pools: Array<V4SubgraphPool | V3SubgraphPool | V2SubgraphPool>
+): Array<V4SubgraphPool> {
+  const v4PoolsByTokenPairsAndFees: Record<string, Array<V4SubgraphPool>> = {}
+
+  ;(pools as Array<V4SubgraphPool>).forEach((pool: V4SubgraphPool) => {
+    if (routableHooks(pool)) {
+      const v4Pools = v4PoolsByTokenPairsAndFees[pool.token0.id.concat(pool.token1.id).concat(pool.feeTier)] || []
+
+      v4Pools.push(pool)
+      v4Pools.sort((a, b) => (a.tvlETH > b.tvlETH ? -1 : 1))
```
Comment (xrsv) [inline test/utils/v4HooksPoolsFiltering.ts:29]: wondering if we want to push/sort/pop on each iteration?
How many pools do we expect to filter? 
also a unit test to validate we get the top 10 would be nice here.

File: test/utils/v4HooksPoolsFiltering.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  })
+
+  const topTvlPools = Object.values(v4PoolsByTokenPairsAndFees).reduce((acc, pools) => {
+    return acc.concat(pools)
+  })
+
+  const allowlistedHooksPools = (pools as Array<V4SubgraphPool>).filter((pool: V4SubgraphPool) => {
+    const shouldFilterOut = !HOOKS_ADDRESSES_ALLOWLIST[chainId].includes(pool.hooks.toLowerCase())
+
+    return !shouldFilterOut
```
Comment (xrsv) [inline test/utils/v4HooksPoolsFiltering.ts:46]: this logic here is a bit confusing: you can just `return HOOKS_ADDRESSES_ALLOWLIST[chainId].includes(pool.hooks.toLowerCase())`

File: test/utils/v4HooksPoolsFiltering.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import { V2SubgraphPool, V3SubgraphPool, V4SubgraphPool } from '@uniswap/smart-order-router'
```
Comment (xrsv) [inline test/utils/v4HooksPoolsFiltering.ts:1]: this file shouldn't be under `test` folder

File: test/utils/v4HooksPoolsFiltering.ts [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+import { Hook } from '@uniswap/v4-sdk'
+import { HOOKS_ADDRESSES_ALLOWLIST } from '../../lib/util/hooksAddressesAllowlist'
+import { ChainId } from '@uniswap/sdk-core'
+
+export function routableHooks(pool: V4SubgraphPool): boolean {
```
Comment (xrsv) [inline test/utils/v4HooksPoolsFiltering.ts:6]: would rename this to 'isHookPoolRoutable(...)`

**Review by xrsv — COMMENTED** [review]
thanks - left some comments

File: test/utils/v4HooksPoolsFiltering.ts (hunk 3 above)
Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts:29]: can prob do an insertion sort

Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts:29]: or priority queue

File: test/utils/v4HooksPoolsFiltering.ts (hunk 2 above)
Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts]: yep https://github.com/Uniswap/routing-api/pull/916/files#diff-86f50fb324e299fbcd9437c599e1b23973141b1325152c8077e9d5dabf30aa1aR9

File: test/utils/v4HooksPoolsFiltering.ts (hunk 3 above)
Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts]: decided to use a priority queue library

File: test/utils/v4HooksPoolsFiltering.ts (hunk 6 above)
Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts]: yes

File: test/utils/v4HooksPoolsFiltering.ts (hunk 4 above)
Comment (jsy1218) [reply test/utils/v4HooksPoolsFiltering.ts]: yep

**Review by xrsv — APPROVED** [review]
nice

**Comment by jsy1218** [comment]
### Merge activity

* **Nov 15, 6:59 PM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/Uniswap/routing-api/916).

**Pull request merged**